File: packages/functions/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/functions/FirebaseFunctions.java
package io.capawesome.capacitorjs.plugins.firebase.functions;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.capawesome.capacitorjs.plugins.firebase.functions.classes.options.CallByNameOptions;
import io.capawesome.capacitorjs.plugins.firebase.functions.classes.options.CallByUrlOptions;
import io.capawesome.capacitorjs.plugins.firebase.functions.classes.results.CallResult;
import io.capawesome.capacitorjs.plugins.firebase.functions.interfaces.NonEmptyResultCallback;
import java.net.URL;
import java.util.concurrent.TimeUnit;

public class FirebaseFunctions {

    private FirebaseFunctionsPlugin plugin;

    public FirebaseFunctions(FirebaseFunctionsPlugin plugin) {
        this.plugin = plugin;
    }

    public void callByName(@NonNull CallByNameOptions options, @NonNull NonEmptyResultCallback callback) {
        String name = options.getName();
        String region = options.getRegion();
        Object data = options.getData();
        Long timeout = options.getTimeout();

        com.google.firebase.functions.HttpsCallableReference callable = getFirebaseFunctionsInstance(region).getHttpsCallable(name);

        if (timeout != null) {
            callable.setTimeout(timeout, TimeUnit.MILLISECONDS);
        }

        callable
            .call(data)
            .addOnSuccessListener(task -> {
                CallResult result = new CallResult(task.getData());
                callback.success(result);
            })
            .addOnFailureListener(exception -> {
                callback.error(exception);
            });
    }

    public void callByUrl(@NonNull CallByUrlOptions options, @NonNull NonEmptyResultCallback callback) {
        URL url = options.getUrl();
        Object data = options.getData();
        Long timeout = options.getTimeout();

        com.google.firebase.functions.HttpsCallableReference callable = getFirebaseFunctionsInstance(null).getHttpsCallableFromUrl(url);

        if (timeout != null) {
            callable.setTimeout(timeout, TimeUnit.MILLISECONDS);
        }

        callable
            .call(data)
            .addOnSuccessListener(task -> {
                CallResult result = new CallResult(task.getData());
                callback.success(result);
            })
            .addOnFailureListener(exception -> {
                callback.error(exception);
            });
    }

    public void useEmulator(@NonNull String host, int port) {
        getFirebaseFunctionsInstance(null).useEmulator(host, port);
    }

    private com.google.firebase.functions.FirebaseFunctions getFirebaseFunctionsInstance(@Nullable String regionOrCustomDomain) {
        if (regionOrCustomDomain == null) {
            return com.google.firebase.functions.FirebaseFunctions.getInstance();
        } else {
            return com.google.firebase.functions.FirebaseFunctions.getInstance(regionOrCustomDomain);
        }
    }
}

File: packages/functions/ios/Plugin/FirebaseFunctions.swift
import Foundation
import FirebaseCore
import FirebaseFunctions

@objc public class FirebaseFunctions: NSObject {
    private let plugin: FirebaseFunctionsPlugin

    init(plugin: FirebaseFunctionsPlugin) {
        self.plugin = plugin
        super.init()
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
    }

    @objc public func callByName(_ options: CallByNameOptions, completion: @escaping (Result?, Error?) -> Void) {
        let name = options.getName()
        let data = options.getData()
        let timeout = options.getTimeout()

        let functions = getFunctions(options.getRegion())
        let callable = functions.httpsCallable(name)
        if let timeout = timeout {
            callable.timeoutInterval = TimeInterval(timeout) / 1000.0
        }
        callable.call(data) { (result, error) in
            if let error = error {
                completion(nil, error)
            } else {
                let result = CallResult(data: result?.data)
                completion(result, nil)
            }
        }
    }

    @objc public func callByUrl(_ options: CallByUrlOptions, completion: @escaping (Result?, Error?) -> Void) {
        let url = options.getUrl()
        let data = options.getData()
        let timeout = options.getTimeout()

        let functions = getFunctions(nil)
        let callable = functions.httpsCallable(url)
        if let timeout = timeout {
            callable.timeoutInterval = TimeInterval(timeout) / 1000.0
        }
        callable.call(data) { (result, error) in
            if let error = error {
                completion(nil, error)
            } else {
                let result = CallResult(data: result?.data)
                completion(result, nil)
            }
        }
    }

    @objc func useEmulator(_ host: String, _ port: Int, _ regionOrCustomDomain: String?) {
        let functions = getFunctions(regionOrCustomDomain)
        functions.useEmulator(withHost: host, port: port)
    }

    private func getFunctions(_ regionOrCustomDomain: String?) -> Functions {
        guard let value = regionOrCustomDomain else {
            return Functions.functions()
        }
        if value.hasPrefix("http://") || value.hasPrefix("https://") {
            return Functions.functions(customDomain: value)
        }
        return Functions.functions(region: value)
    }
}
